### PR TITLE
Android native video

### PR DESCRIPTION
**Pull request by HeikkiSistonen**
Android native video capture activity. 

-At the moment it takes only max video length as a parameter.
-Activity is launched from javascript/typescript code and it gives url of the recorded video in callback parameter. 
-When user cancel no result is given. Activity is finished.

### CODE REVIEW

File: src/android/AutoFitTextureView.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+     * Sets the aspect ratio for this view. The size of the view will be measured based on the ratio
+     * calculated from the parameters. Note that the actual sizes of parameters don't matter, that
+     * is, calling setAspectRatio(2, 3) and setAspectRatio(4, 6) make the same result.
+     *
+     * @param width Relative horizontal size
+     * @param height Relative vertical size
+     */
+    public void setAspectRatio(int width, int height) {
+        if (width < 0 || height < 0) {
+            throw new IllegalArgumentException("Size cannot be negative.");
```
Comment (KamilMakowski) [inline src/android/AutoFitTextureView.java:29]: can ratio be of value 0?

File: src/android/AutoFitTextureView.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        int width = MeasureSpec.getSize(widthMeasureSpec);
+        int height = MeasureSpec.getSize(heightMeasureSpec);
+        if (0 == mRatioWidth || 0 == mRatioHeight) {
+            setMeasuredDimension(width, height);
+        } else {
+            if (width < height * mRatioWidth / mRatioHeight) {
+                setMeasuredDimension(width, width * mRatioHeight / mRatioWidth);
+            } else {
+                setMeasuredDimension(height * mRatioWidth / mRatioHeight, height);
+            }
```
Comment (KamilMakowski) [inline src/android/AutoFitTextureView.java:47]: what does that magic calculation do? Setting horizontal or vertical view? Maybe consider some comment here

File: src/android/MainActivity.java [hunk 3]
```diff
@@ -0,0 +1,652 @@
...
+                @Override
+                public void onConfigured( CameraCaptureSession cameraCaptureSession) {
+                    previewSession = cameraCaptureSession;
+                    updatePreview();
+
+                    MainActivity.this.runOnUiThread(() -> {
+                        isRecordingVideo = true;
+                        // Start recording
+                        recorder.start();
+                        lockDeviceRotation(true);
```
Comment (KamilMakowski) [inline src/android/MainActivity.java]: idk if that matter but should we first lock rotation and then start recording. meaning moving this line one up

File: src/android/MediaCapture.java [hunk 4]
```diff
@@ -204,48 +71,55 @@ public void run() {
...
         }
-        return flashAvailable;
+        // Use internal storage
+        else {
+            cache = cordova.getActivity().getCacheDir();
+        }
+
+        // Create the cache directory if it doesn't exist
+        cache.mkdirs();
+        return cache.getAbsolutePath();
```
Comment (KamilMakowski) [inline src/android/MediaCapture.java:107]: are those files clear in situations of:

- [ ] video file successfully recorded. After it is moved to app directory, is the original file deleted?
- [ ] video file recording started but not finished. application crashed/killed by user or system. is the original file deleted, after relunch of app, or this java component?

**Review by KamilMakowski — COMMENTED** [review]
Vertical and horizontal recording is supported?
If video has been recorded horizontally, preview of recorded video is in the correct shape in the application?

File: src/android/MainActivity.java [hunk 5]
```diff
@@ -0,0 +1,656 @@
...
+                    break;
+                case SENSOR_ORIENTATION_INVERSE_DEGREES:
+                    recorder.setOrientationHint(INVERSE_ORIENTATIONS.get(rotation));
+                    break;
+            }
+
+            recorder.prepare();
+
+        } catch (IOException e) {
+            e.printStackTrace();
```
Comment (mamipu) [inline src/android/MainActivity.java:268]: Should this be also thrown? Now it let's IOExceptions pass

File: src/android/MainActivity.java [hunk 6]
```diff
@@ -0,0 +1,656 @@
...
+                        }, 0, 1000);
+                    });
+
+                }
+                @Override
+                public void onConfigureFailed( CameraCaptureSession cameraCaptureSession) {
+                    Log.e("", "onConfigureFailed: Failed");
+                }
+            }, backgroundHandler);
+        } catch (CameraAccessException e) {
```
Comment (mamipu) [inline src/android/MainActivity.java:344]: If camera access is not accepted, what should happen? Probably not a problem to be solved right now

File: src/android/MainActivity.java [hunk 7]
```diff
@@ -0,0 +1,656 @@
...
+                }
+                configureTransform(width, height);
+                recorder = new MediaRecorder();
+
+                manager.openCamera(camerId, stateCallback, null);
+            } catch (CameraAccessException e) {
+                e.printStackTrace();
+            }
+        } catch (Exception e) {
+            Log.d("MyCameraApp", e.getMessage());
```
Comment (mamipu) [inline src/android/MainActivity.java:467]: All exceptions from openCamera are only logged, probably should throw error, maybe remove the try catch or add the throw?

File: src/android/MediaCapture.java [hunk 8]
```diff
@@ -204,48 +71,55 @@ public void run() {
...
-                if (PackageManager.FEATURE_CAMERA_FLASH.equalsIgnoreCase(feature.name)) {
-                    flashAvailable = true;
-                    break;
-                }
-            }
+    private String getTempDirectoryPath() {
+        File cache = null;
+
+        // SD Card Mounted
+        if (Environment.getExternalStorageState().equals(Environment.MEDIA_MOUNTED)) {
```
Comment (mamipu) [inline src/android/MediaCapture.java:97]: SD card situation does it actually work when reading the file and deleting the file from storage?

**Review by mamipu — APPROVED** [review]

File: src/android/MainActivity.java (hunk 3 above)
Comment (HeikkiSistonen) [reply src/android/MainActivity.java]: fixed

File: src/android/MediaCapture.java (hunk 4 above)
Comment (HeikkiSistonen) [reply src/android/MediaCapture.java:107]: fixed

**Review by KamilMakowski — APPROVED** [review]

File: src/android/MainActivity.java (hunk 6 above)
Comment (HeikkiSistonen) [reply src/android/MainActivity.java:344]: to be solved later

File: src/android/MainActivity.java (hunk 7 above)
Comment (HeikkiSistonen) [reply src/android/MainActivity.java:467]: error are thrown

File: src/android/MediaCapture.java (hunk 8 above)
Comment (HeikkiSistonen) [reply src/android/MediaCapture.java:97]: to be fixed later if needed

**Pull request merged**